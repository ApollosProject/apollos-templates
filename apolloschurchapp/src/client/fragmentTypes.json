{"__schema":{"types":[{"kind":"INTERFACE","name":"Node","possibleTypes":[{"name":"ContentChannel"},{"name":"Person"},{"name":"Campus"},{"name":"Event"},{"name":"Group"},{"name":"Scripture"},{"name":"Pass"},{"name":"UniversalContentItem"},{"name":"DevotionalContentItem"},{"name":"MediaContentItem"},{"name":"ContentSeriesContentItem"},{"name":"WeekendContentItem"},{"name":"ActionListFeature"},{"name":"TextFeature"},{"name":"ScriptureFeature"}]},{"kind":"INTERFACE","name":"ContentItem","possibleTypes":[{"name":"UniversalContentItem"},{"name":"DevotionalContentItem"},{"name":"MediaContentItem"},{"name":"ContentSeriesContentItem"},{"name":"WeekendContentItem"}]},{"kind":"INTERFACE","name":"Media","possibleTypes":[{"name":"ImageMedia"},{"name":"VideoMedia"},{"name":"AudioMedia"}]},{"kind":"INTERFACE","name":"MediaSource","possibleTypes":[{"name":"ImageMediaSource"},{"name":"VideoMediaSource"},{"name":"AudioMediaSource"}]},{"kind":"INTERFACE","name":"Sharable","possibleTypes":[{"name":"SharableContentItem"},{"name":"SharableFeature"}]},{"kind":"INTERFACE","name":"Feature","possibleTypes":[{"name":"ActionListFeature"},{"name":"TextFeature"},{"name":"ScriptureFeature"}]}]}}