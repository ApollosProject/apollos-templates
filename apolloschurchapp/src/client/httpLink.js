import { Platform } from 'react-native';
import { createUploadLink } from 'apollo-upload-client';
import ApollosConfig from '@apollosproject/config';
import { split } from 'apollo-link';
import { createHttpLink } from 'apollo-link-http';
import { getMainDefinition } from 'apollo-utilities';

let uri = ApollosConfig.APP_DATA_URL;
const androidUri = ApollosConfig.ANDROID_URL || '10.0.2.2';

// Android's emulator requires localhost network traffic to go through 10.0.2.2
if (Platform.OS === 'android') uri = uri.replace('localhost', androidUri);

export default split(
  ({ query }) => {
    const { kind, operation } = getMainDefinition(query);
    return kind === 'OperationDefinition' && operation === 'mutation';
  },
  createUploadLink({ uri }),
  createHttpLink({
    uri,
    useGETForQueries: true,
    headers: {
      // We can safely send these headers.
      // Fastly does not currently respect no-store or no-cache directives. Including either or both of these in a Cache-Control header has no effect on Fastly's caching decision
      // https://docs.fastly.com/en/guides/configuring-caching#do-not-cache
      'Cache-Control': 'no-cache, no-store',
    },
  })
);
