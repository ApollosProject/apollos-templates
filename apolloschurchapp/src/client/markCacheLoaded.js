import gql from 'graphql-tag';

export default gql`
  mutation markCacheLoaded {
    cacheMarkLoaded @client
  }
`;
