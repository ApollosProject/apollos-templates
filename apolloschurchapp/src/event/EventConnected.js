import React, { PureComponent } from 'react';
import { Query } from '@apollo/client/react/components';
import { get } from 'lodash';
import PropTypes from 'prop-types';

import { ErrorCard, ThemeMixin } from '@apollosproject/ui-kit';

import { TrackEventWhenLoaded } from '@apollosproject/ui-analytics';

import Event from './Event';
import GET_EVENT from './getEvent';

class EventConnected extends PureComponent {
  static propTypes = {
    navigation: PropTypes.shape({
      push: PropTypes.func,
    }),
    route: PropTypes.shape({
      params: PropTypes.shape({
        eventId: PropTypes.string,
      }),
    }),
  };

  get eventId() {
    return this.props.route.params.eventId;
  }

  get queryVariables() {
    return { eventId: this.eventId };
  }

  renderWithData = ({ loading, error, data }) => {
    if (error) return <ErrorCard error={error} />;

    const event = data?.node || {};

    const { theme = {} } = event;

    return (
      <ThemeMixin
        mixin={{
          type: get(theme, 'type', 'light').toLowerCase(),
          colors: get(theme, 'colors'),
        }}
      >
        <TrackEventWhenLoaded
          loaded={!!(!loading && event.name)}
          eventName={'View Event'}
          properties={{
            title: event.name,
            eventId: this.eventId,
          }}
        />
        <Event
          id={this.eventId}
          event={event}
          loading={loading}
          error={error}
        />
      </ThemeMixin>
    );
  };

  render() {
    return (
      <Query
        query={GET_EVENT}
        fetchPolicy="network-only"
        variables={this.queryVariables}
      >
        {this.renderWithData}
      </Query>
    );
  }
}

export default EventConnected;
