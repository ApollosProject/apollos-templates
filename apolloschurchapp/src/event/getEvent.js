import gql from 'graphql-tag';

export const EVENT_FRAGMENT = gql`
  fragment eventFragment on Event {
    id
    name
    location
    start
    end
    description
    image {
      sources {
        uri
      }
    }
  }
`;

export default gql`
  query getEvent($eventId: ID!) {
    node(id: $eventId) {
      __typename
      ... on Event {
        ...eventFragment
      }
    }
  }
  ${EVENT_FRAGMENT}
`;
