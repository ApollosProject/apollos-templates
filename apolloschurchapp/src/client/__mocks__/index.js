import React from 'react';
import { MockedProvider } from 'react-apollo/test-utils';
import ApolloClient from 'apollo-client';
import { makeExecutableSchema, addMockFunctionsToSchema } from 'graphql-tools';
import { SchemaLink } from 'apollo-link-schema';
import { importSchema } from 'graphql-import';

import cache from '../cache';
import { resolvers, schema, defaults } from '../../store';

// eslint-disable-next-line
export default (props) => {
  let finalProps = props;
  // eslint-disable-next-line
  if (!props.mocks) {
    // eslint-disable-next-line
    finalProps = { ...props, resolvers };
  }
  return <MockedProvider cache={cache} {...finalProps} />;
};

const serverSchema = makeExecutableSchema({
  typeDefs: importSchema(
    `${process.cwd()}/../apollos-church-api/local.graphql`
  ),
  resolverValidationOptions: {
    requireResolversForResolveType: false,
  },
});
addMockFunctionsToSchema({ schema: serverSchema });

const link = new SchemaLink({ schema: serverSchema });
cache.writeData({ data: defaults });

export const client = new ApolloClient({
  link,
  cache,
  resolvers,
  typeDefs: schema,
});
