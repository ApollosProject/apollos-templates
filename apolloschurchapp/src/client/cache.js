import { InMemoryCache } from '@apollo/client/cache';
import AsyncStorage from '@react-native-community/async-storage';
import { CachePersistor } from 'apollo3-cache-persist';
import ApollosConfig from '@apollosproject/config';

import introspectionQueryResultData from './fragmentTypes.json';

const finalPossibleTypes = {};
introspectionQueryResultData.__schema.types.forEach((supertype) => {
  if (supertype.possibleTypes) {
    finalPossibleTypes[supertype.name] = [
      ...supertype.possibleTypes.map((subtype) => subtype.name),
    ];
  }
});
// We reset our apollo cache based an env value and static number.
// In the future, we should also look at resetting the app when an error occurs related to Apollo.
// You can also increment this number to force a manual reset of the cache.
const SCHEMA_VERSION = `${ApollosConfig.SCHEMA_VERSION}-1`; // Must be a string.
const SCHEMA_VERSION_KEY = 'apollo-schema-version';

const nodeCacheRedirect = (_, { id }, { getCacheKey }) =>
  id ? getCacheKey({ __typename: id.split(':')[0], id }) : null;

const cache = new InMemoryCache({
  possibleTypes: finalPossibleTypes,
  cacheRedirects: {
    Query: {
      node: nodeCacheRedirect,
    },
  },
});

const persistor = new CachePersistor({
  cache,
  storage: AsyncStorage,
});

// 1. If we are on the same schema version, restore the cache.
// 2. If that fails, purge the cache and update the stored version so we don't try and restore agian
// 3. If we are on a new schema version, purge the cache.
// 4. If purging or setting AsyncStorage keys fails (or anything else) we capture the error and log it.

export const ensureCacheHydration = (async () => {
  // We wrap everything in try/catch because crashing on a cache restore is bad
  // ðŸ˜¬
  try {
    const currentVersion = await AsyncStorage.getItem(SCHEMA_VERSION_KEY);
    if (currentVersion === SCHEMA_VERSION) {
      // If the current version matches the latest version,
      // we're good to go and can restore the cache.
      try {
        await persistor.restore();
      } catch (restoreError) {
        // If the restore fails, we want to do our best to purge the cache.
        await persistor.purge();
        console.error('Error restoring cache, purging the cache', restoreError);
      }
    } else {
      // Otherwise, we'll want to purge the outdated persisted cache
      // and mark ourselves as having updated to the latest version.
      await persistor.purge();
      await AsyncStorage.setItem(SCHEMA_VERSION_KEY, SCHEMA_VERSION);
    }
  } catch (error) {
    console.error('Error restoring or purging Apollo cache', error);
  }
})();

export default cache;
