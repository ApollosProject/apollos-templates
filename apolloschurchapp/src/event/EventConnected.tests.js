import React from 'react';

import moment from 'moment';
import { MockedProvider } from '@apollo/client/testing';
import { Providers, renderWithApolloData } from '@apollosproject/ui-test-utils';
import getEvent from './getEvent';
import EventConnected from './EventConnected';

const mocks = {
  request: {
    query: getEvent,
    variables: { eventId: 'Event:123' },
  },
  result: {
    data: {
      node: {
        __typename: 'Event',
        id: 'Event:123',
        start: moment('2019-09-26T15:10:51.200Z')
          .utc()
          .toJSON(),
        end: moment('2019-09-26T17:10:51.200Z')
          .utc()
          .toJSON(),
        location: 'Willow Creek, Chicago',
        name: 'Cookout',
        description: 'Some detailed description',
        image: {
          __typename: 'ImageMedia',
          sources: [
            {
              uri: 'https://url.com/image.jpg',
              __typename: 'ImageMediaSource',
            },
          ],
        },
      },
    },
  },
};

describe('EventConnected component', () => {
  it('renders without errors', async () => {
    const tree = await renderWithApolloData(
      <Providers mocks={[mocks]} MockedProvider={MockedProvider}>
        <EventConnected route={{ params: { eventId: 'Event:123' } }} />
      </Providers>
    );

    expect(tree).toMatchSnapshot();
  });
});
