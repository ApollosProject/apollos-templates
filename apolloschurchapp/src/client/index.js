import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { ApolloLink } from 'apollo-link';
import { getVersion, getApplicationName } from 'react-native-device-info';

import { authLink, buildErrorLink } from '@apollosproject/ui-auth';
import { resolvers, schema, defaults } from '../store';
import NavigationService from '../NavigationService';
import httpLink from './httpLink';
import cache, { ensureCacheHydration, MARK_CACHE_LOADED } from './cache';

const goToAuth = () => NavigationService.resetToAuth();
const wipeData = () => cache.writeData({ data: defaults });

let clearStore;
let storeIsResetting = false;
const onAuthError = async () => {
  if (!storeIsResetting) {
    storeIsResetting = true;
    await clearStore();
  }
  storeIsResetting = false;
  goToAuth();
};

const errorLink = buildErrorLink(onAuthError);

const link = ApolloLink.from([authLink, errorLink, httpLink]);

export const client = new ApolloClient({
  link,
  cache,
  queryDeduplication: false,
  shouldBatch: true,
  resolvers,
  typeDefs: schema,
  name: getApplicationName(),
  version: getVersion(),
});

// Hack to give auth link access to method on client;
// eslint-disable-next-line prefer-destructuring
clearStore = client.clearStore;

wipeData();
// Ensure that media player still works after logout.
client.onClearStore(() => wipeData());

class ClientProvider extends PureComponent {
  static propTypes = {
    client: PropTypes.shape({
      cache: PropTypes.shape({}),
    }),
  };

  static defaultProps = {
    client,
  };

  async componentDidMount() {
    try {
      await ensureCacheHydration;
    } catch (e) {
      throw e;
    } finally {
      client.mutate({ mutation: MARK_CACHE_LOADED });
    }
  }

  render() {
    return <ApolloProvider {...this.props} client={client} />;
  }
}

export default ClientProvider;
