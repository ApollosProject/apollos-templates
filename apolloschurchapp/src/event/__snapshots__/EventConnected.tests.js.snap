// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventConnected component renders without errors 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;
